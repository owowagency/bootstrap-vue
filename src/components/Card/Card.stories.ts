import Card from '.';

export default {
    title: 'Components/Cards',
    component: Card,
    argTypes: {},
};

export const Example = (args) => ({
    components: {Card},
    setup() {
        return {args};
    },
    template: `
        <Card
            no-body
            style="width: 18rem"
            v-bind="args"
        >
            <img src="https://picsum.photos/id/237/200/130" class="card-img-top" alt="aaa">

            <div class="card-body">
                <h5 class="card-title">Card title</h5>
                <p class="card-text">Some quick example text to build on the card title and make up the bulk of the card's content.</p>
                <a href="#" class="btn btn-primary">Go somewhere</a>
            </div>
        </Card>
    `,
});

export const Body = (args) => ({
    components: {Card},
    setup() {
        return {args};
    },
    template: `
        <Card
            no-body
            v-bind="args"
        >
            <div class="card-body">
                I am a body
            </div>

            <div class="card-body">
                I am a body
            </div>
        </Card>
    `,
});

export const TitlesTextLinks = (args) => ({
    components: {Card},
    setup() {
        return {args};
    },
    template: `
        <Card v-bind="args">
            <h5 class="card-title">Card title</h5>
            <h6 class="card-subtitle mb-2">Card subtitle</h6>
            <p class="card-text">Some quick example text to build on the card title and make up the bulk of the card's content.</p>
            <a href="#" class="card-link">Card link</a>
            <a href="#" class="card-link">Another link</a>
        </Card>
    `,
});

export const HeaderFooter = (args) => ({
    components: {Card},
    setup() {
        return {args};
    },
    template: `
        <Card
            footer="I am a footer"
            header="I am a header"
            text="I am just normal text"
            v-bind="args"
        />
    `,
});
