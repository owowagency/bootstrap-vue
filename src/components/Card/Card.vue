<template>
    <div class="card">
        <div
            v-if="!!$slots.header || !!header"
            class="card-header"
        >
            <!-- @slot Displays content inside the header of the card -->
            <slot name="header">
                {{ header }}
            </slot>
        </div>

        <div :class="{'card-body': !noBody}">
            <!-- @slot Displays content inside the body of the card -->
            <slot name="default">
                <h5
                    v-if="title"
                    class="card-title"
                >
                    {{ title }}
                </h5>

                <h6
                    v-if="subtitle"
                    class="card-subtitle"
                >
                    {{ subtitle }}
                </h6>

                <div
                    v-if="text"
                    class="card-text"
                >
                    {{ text }}
                </div>
            </slot>
        </div>

        <div
            v-if="!!$slots.footer || footer"
            class="card-footer"
        >
            <!-- @slot Displays content inside the footer of the card -->
            <slot name="footer">
                {{ footer }}
            </slot>
        </div>
    </div>
</template>

<script lang="ts" setup>
defineProps({
    /**
     * The text displayed in the footer. Will be overwritten by the user of the `footer` slot
     */
    footer: {
        type: String,
        default: undefined,
    },
    /**
     * The text displayed in the header. Will be overwritten by the user of the `header` slot
     */
    header: {
        type: String,
        default: undefined,
    },
    /**
     * Prevents the `.card-body` class from being added. Useful for when multiple card bodies are required
     */
    noBody: {
        type: Boolean,
        default: false,
    },
    /**
     * The subtitle displayed in the body. Will be overwritten by the user of the `default` slot
     */
    subtitle: {
        type: String,
        default: undefined,
    },
    /**
     * The text displayed in the body. Will be overwritten by the user of the `default` slot
     */
    text: {
        type: String,
        default: undefined,
    },
    /**
     * The title displayed in the body. Will be overwritten by the user of the `default` slot
     */
    title: {
        type: String,
        default: undefined,
    },
});
</script>

<docs>
```vue
<Card
    subtitle="Card subtitle"
    text="Some quick example text to build on the card title and make up the bulk of the card's content."
    title="Card title"
/>

<Card no-body>
    <template #default>
        <div class="card-body">
            I have ...
        </div>

        <div class="card-body">
            ... two bodies.
        </div>
    </template>

    <template #footer>
        <Button>
            That is great!
        </Button>
    </template>
</Card>
```
</docs>
