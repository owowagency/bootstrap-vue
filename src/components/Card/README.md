# Card

```vue
<Card
  subtitle="Card subtitle"
  text="Some quick example text to build on the card title and make up the bulk of the card's content."
  title="Card title"
/>

<Card no-body>
    <template #default>
        <div class="card-body">
            I have ...
        </div>

        <div class="card-body">
            ... two bodies.
        </div>
    </template>

    <template #footer>
        <Button>
            That is great!
        </Button>
    </template>
</Card>
```

## Props

| Prop name | Description                                                                                         | Type    | Values | Default   |
| --------- | --------------------------------------------------------------------------------------------------- | ------- | ------ | --------- |
| footer    | The text displayed in the footer. Will be overwritten by the user of the `footer` slot              | string  | -      | undefined |
| header    | The text displayed in the header. Will be overwritten by the user of the `header` slot              | string  | -      | undefined |
| noBody    | Prevents the `.card-body` class from being added. Useful for when multiple card bodies are required | boolean | -      | false     |
| subtitle  | The subtitle displayed in the body. Will be overwritten by the user of the `default` slot           | string  | -      | undefined |
| text      | The text displayed in the body. Will be overwritten by the user of the `default` slot               | string  | -      | undefined |
| title     | The title displayed in the body. Will be overwritten by the user of the `default` slot              | string  | -      | undefined |

## Slots

| Name    | Description                                    | Bindings |
| ------- | ---------------------------------------------- | -------- |
| header  | Displays content inside the header of the card |          |
| default | Displays content inside the body of the card   |          |
| footer  | Displays content inside the footer of the card |          |
