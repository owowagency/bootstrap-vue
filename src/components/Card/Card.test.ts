import {describe, expect, it} from 'vitest';
import Card from '.';
import {shallowMount} from '@vue/test-utils';

const textPropTest = (propName: string) => {
    it(`renders ${propName}`, () => {
        const text = `I am a ${propName}`;

        const wrapper = shallowMount(Card, {
            props: {[propName]: text},
        });

        expect(wrapper.find(`.card-${propName}`).text()).toBe(text);
    });
};

describe('template', () => {
    componentRenderTest(Card);

    componentSlotRenderTest(Card, 'header');

    componentSlotRenderTest(Card);

    componentSlotRenderTest(Card, 'footer');

    ['footer', 'header', 'subtitle', 'text', 'title'].forEach(textPropTest);

    it('does not add body class', () => {
        const wrapper = shallowMount(Card, {
            props: {noBody: true},
        });

        expect(wrapper.find('.card-body').exists()).toBe(false);
    });
});
