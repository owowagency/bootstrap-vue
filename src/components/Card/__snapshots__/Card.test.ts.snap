// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<div class=\\"card\\">
  <!--v-if-->
  <div class=\\"card-body\\">
    <!-- @slot Displays content inside the body of the card -->
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>"
`;
